kong/plugins/oauth2/access.lua

```diff
@@ -1045,20 +1045,20 @@ function _M.execute(conf)
     return
   end
 
-  if kong.request.get_method() == "POST" then
```

Comment:
The spec says:

Token Endpoint (https://tools.ietf.org/html/rfc6749#section-3.2):

> The client MUST use the HTTP "POST" method when making access token requests.

Authorization Endpoint:
>  The authorization server MUST support the use of the HTTP "GET" method [RFC2616] for the authorization endpoint and MAY support the use of the "POST" method as well.


Thus can we change it so that `POST` is checked for `token endpoint` and `GET/POST` is checked for `authorization endpoint`?

Reply:
The issue in my mind is Kong is placing these on proxy routes that also serve the service via reverse proxy so its not a standalone authorization server. By checking HTTP Methods and certain portions of pathing for things like ```/oauth2/*``` I think it causes some behavior that would not be what the client should get.

Example is when you send a GET to ```/oauth2/token``` the client should get a 400 bad request(Because they didn't send a POST) but because of the fall though behavior it will hit, the client will get a 401(not authorized, which isn't the correct status when the wrong method was sent). 

Are you suggesting to add the HTTP checks if the string match is found and if HTTP != POST throw a 400(thus not hitting the 401 incorrect status?) because then it covers both cases, I mean like this seems correct behavior you could achieve from a client perspective to me:

```
  local from = string_find(path, "/oauth2/token", nil, true)
    if from then
      if kong.request.get_method() == "POST" then
        return authorize(conf)	    return issue_token(conf)
     else
       {return 400 bad request}
    end	
  end

  from = string_find(path, "/oauth2/authorize", nil, true)
  if from then
      if kong.request.get_method() == "POST" or kong.request.get_method() == "GET" then
        return authorize(conf)
     else
       {return 400 bad request}
  end
```

Because if the routing URL has ```/oauth2/token``` or ```/oauth2/authorize``` the client is certainly indending to perform OAuth2.0 logic and not trying to hit the proxy without a bearer token, yeilding the 401's you would get today by sending a GET.

Reply:
cc @bungle ^